### PR TITLE
Travis Style Check Support

### PR DESCRIPTION
**Pull request by cbaenziger**
This provides for us to enable [Travis CI](http://travis-ci.org) for pull request checking. Particularly this adds support for us to start using `rake` along with a few tasks:

* `rake clean` -- A stub to clean the repository up after a cluster build
* `rake style` -- Runs all `style:*` verbos checks
* `rake style:diff` -- Provides if your repo improves or degrades style (a quick go/no-go one can run)
* `rake style:foodcritic` -- foodcritic for all cookbooks
* `rake style:rubocop` -- RuboCop

I have tested this with TravisCI ensuring we improve or at least maintain style quality every commit. As soon as this goes into master, I can work with the Bloomberg organization owners to enable chef-bach for TravisCI.

### CODE REVIEW

**Comment by cbaenziger** [comment]
For reference, I had to implement dynamic foodcritic style targets per cookbook due to issues whereby I could not get the FoodCritic raketask to effectively run `foodcritic -E stub-environment/environments/Test-Laptop.json -R stub-environment/roles -c 12.0.0 cookbooks/*`. There is some discussion of this at foodcritic/foodcritic#250 for future following.

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,18 @@
+language: ruby
+sudo: false
+rvm:
+  - 2.0
+  - 2
+bundler_args: --without cookbooks
+cache: bundler
+gemfile: Gemfile
+before_install:
+  - chef --version &> /dev/null || curl -L https://www.getchef.com/chef/install.sh | sudo bash -s -- -P chefdk
```
Comment (aespinosa) [inline .travis.yml]: should we pin this to what is set to https://github.com/bloomberg/chef-bach/blob/master/build_bins.sh#L25 ?

File: Rakefile [hunk 2]
```diff
@@ -0,0 +1,70 @@
+#!/usr/bin/env rake
+
+chef_version = '12.14.89'
```
Comment (aespinosa) [inline Rakefile]: I think we're at 12.19.36

File: Rakefile [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+  desc 'Run foodcritic checks'
+  task chef: %w(style:chef./) + \
+              Dir.glob('./cookbooks/*').map { |c| 'style:chef' + c }
+end
+
+desc 'Run style checks'
+task style: %w(style:ruby style:chef)
+
+desc 'Clean some generated files'
```
Comment (aespinosa) [inline Rakefile]: Are we planning to have other non-file cleaning related stuff? If not let's use the existing clean/ clobber facilities: http://docs.seattlerb.org/rake/doc/rakefile_rdoc.html#label-Clean+and+Clobber+Tasks

Comment (cbaenziger) [reply Rakefile]: Thanks for this pointer; mind if I hold this suggestion until I start the `rspec`/`chefspec` work where I will need to make sure Gem caches get cleaned-up?

File: .travis.yml (hunk 1 above)
Comment (cbaenziger) [reply .travis.yml]: Fixed to use the same version for now; ideally master should float though.

**Comment by cbaenziger** [comment]
Hi @aespinosa let me know if this looks better and I can hold off the Rake changes for the next update. I verified this ran through on my personal Travis build test.

File: compare_style.sh [hunk 4]
```diff
@@ -0,0 +1,141 @@
+#!/bin/bash
```
Comment (aespinosa) [inline compare_style.sh]: This shell script seems to be more of a crutch for Travis since it doesn't support tracking build data.

I think codeclimate will be the better public service tracking linting metrics.

Plus we won't need to maintain another shellscript

File: Rakefile [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+chef_version = '12.19.36'
+namespace :style do
+  require 'rubocop/rake_task'
+  desc 'Style check with rubocop'
+
+  require 'foodcritic'
+  desc 'Style check with foodcritic for root'
+  orig_dir = Dir.pwd
+  FoodCritic::Rake::LintTask.new('chef./'.to_sym) do |t|
```
Comment (aespinosa) [inline Rakefile]: Rake tasks around the chefdk ecosystem tends to be less maintained than the main CLI scripts.

Problems in the broader community tent to occur in some versions of ChefDK where each individual tool gets rubygem activation conflicts.

CLI applications work around this by using appbundler where each CLI app can activate its own set of dependencies.  (At least that's how I understood how chefdk works)

TLDR it's probably better to shell out

File: Rakefile [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+        chef_version: chef_version,
+        role_path: ['./stub-environment/roles/*'],
+        cookbook_path: c,
+        environment_path: ['./stub-environments/environments/Test-Laptop.json']
+      }
+    end
+    Dir.chdir(orig_dir)
+  end
+
+  RuboCop::RakeTask.new(:ruby) do |t|
```
Comment (aespinosa) [inline Rakefile]: shellout here as well

**Comment by http-418** [comment]
As discussed in the 1 Feb 18 code review, changes will be needed on this PR:
- Allan's concerns about rake task wrapper objects should be addressed
- The new shell script should be removed

File: compare_style.sh (hunk 4 above)
Comment (cbaenziger) [reply compare_style.sh]: I appreciate being able to check my changes before submitting to an external service but it is optional for one to use. I would rather we have shared tooling than folks implementing their own ad-hoc methods. Indeed CodeClimate seems like a better service than Travis for tracking metrics on code quality though.

File: Rakefile (hunk 5 above)
Comment (cbaenziger) [reply Rakefile]: I have updated to shell out now.

File: Rakefile (hunk 6 above)
Comment (cbaenziger) [reply Rakefile]: Ditto on updating for shell out here.
